File: src/main/java/com/cyhone/classfile/ClassFile.java
package com.cyhone.classfile;

import com.cyhone.exception.ClassFileError;
import com.cyhone.exception.EnvCheckError;
import com.cyhone.types.Uint16;

/**
 * @author cyhone
 * @date 2017/2/20
 */
public class ClassFile {
    private Uint16 minorVersion;
    private Uint16 majorVersion;
    private ConstantPool constantPool;
    private Uint16 accessFlags;
    private Uint16 thisClass;
    private Uint16 superClass;
    private Uint16[] interfaces;
    private MemberInfo[] fields;
    private MemberInfo[] methods;
    private AttributeInfo[] attributes;

    void read(ClassReader reader) {
        readAndCheckMagic(reader);
        readAndCheckVersion(reader);
        this.constantPool = ConstantPool.readConstantPool(reader);
        this.accessFlags = reader.readUint16();
        this.thisClass = reader.readUint16();
        this.superClass = reader.readUint16();
        this.interfaces = reader.readUint16s();
        this.fields = MemberInfo.readMembers(reader, this.constantPool);
        this.methods = MemberInfo.readMembers(reader, this.constantPool);
        this.attributes = AttributeInfo.readAttributes(reader, this.constantPool);
    }

    void readAndCheckMagic(ClassReader reader) {
        int magic = reader.readUint32().toInt();
        if (magic != 0xCAFEBABE) {
            throw new ClassFileError("java.lang.ClassFormatError: magic!");
        }
    }

    void readAndCheckVersion(ClassReader reader) {
        this.minorVersion = reader.readUint16();
        this.majorVersion = reader.readUint16();

        switch (majorVersion.toInt()) {
            case 45:
                return;
            case 46:
            case 47:
            case 48:
            case 49:
            case 50:
            case 51:
            case 52:
                if (minorVersion.toInt() == 0) {
                    return;
                }
        }
        throw new ClassFileError("java.lang.UnsupportedClassVersionError!");
    }
}

File: toyjvm/classfile/classfile.cpp
//
// Created by cyhone on 18-2-8.
//
#include <toyjvm/utilities/bytereader.h>
#include <toyjvm/classfile/classfile.h>
#include <toyjvm/utilities/exception.h>
#include <glog/logging.h>
#include <boost/format.hpp>

using namespace jvm;

#define JAVA_CLASSFILE_MAGIC 0xCAFEBABE
#define JAVA_MIN_SUPPORTED_MAJOR_VERSION 45
#define JAVA_MAX_SUPPORTED_MAJOR_VERSION 52
#define JAVA_MAX_SUPPORTED_MINOR_VERSION 0

ClassFile::~ClassFile()
{
    for (auto field : fields_) {
        delete field;
    }

    for (auto method: methods_) {
        delete method;
    }
}

void ClassFile::parse()
{
    checkMagicAndVersions();
    const_pool_.read(reader_);
    access_flags_ = reader_.read<u2>();
    this_class_ = reader_.read<u2>();
    super_class_ = reader_.read<u2>();
    interfaces_ = reader_.batchRead<u2, u2>();
    fields_ = readMembers<FieldInfo>();
    methods_ = readMembers<MethodInfo>();
    attr_table_.read(reader_, const_pool_);

    //logClassInfo();
}

void ClassFile::logClassInfo()
{
    DLOG(INFO) << "class name : " << const_pool_.classNameOf(this_class_);
    DLOG(INFO) << "super class : " << const_pool_.classNameOf(super_class_);
    DLOG(INFO) << "interfaces : ";
    for (auto inter : interfaces_) {
        DLOG(INFO) << const_pool_.classNameOf(inter);
    }

    DLOG(INFO) << "fields count: " << fields_.size();
    for (auto &f : fields_) {
        DLOG(INFO) << f->memberName();
    }

    DLOG(INFO) << "methods count: " << fields_.size();
    for (auto &m : methods_) {
        DLOG(INFO) << m->memberName();
    }
}


void ClassFile::checkMagicAndVersions()
{
    if (reader_.read<u4>() != JAVA_CLASSFILE_MAGIC) {
        throw ClassFormatError("Magic Number is wrong");
    }

    auto minor_version = reader_.read<u2>();
    auto major_version = reader_.read<u2>();

    if (major_version < JAVA_MIN_SUPPORTED_MAJOR_VERSION
        || major_version > JAVA_MAX_SUPPORTED_MAJOR_VERSION
        || minor_version > JAVA_MAX_SUPPORTED_MINOR_VERSION) {
        throw ClassFormatError(
                (boost::format("Unsupported version: %1%.%2%,")
                 % major_version
                 % minor_version).str()
        );
    }

    DLOG(INFO) << "version: " << major_version << "." << minor_version;
}

